### PR TITLE
Do dockerized builds for minikube and localkube

### PR DESCRIPTION
**Pull request by r2d4**
BUILD_IN_DOCKER=y will build both localkube and minikube in docker
LOCALKUBE_BUILD_IN_DOCKER=y, MINIKUBE_BUILD_IN_DOCKER=y will toggle docker builds for each
respective binary

Regardless of the options provided, if you attempt to build localkube on a
non-linux platform, it will run in docker.  This ensures that `make`
still works as expected on darwin.

The only thing that I couldn't get working was the extra platform detection when cross building.  So BUILD_IN_DOCKER=y is required when `make cross`.  Any suggestions on how to avoid that would be great

I originally planned something like `ifneq(,$(or $(filter $(MINIKUBE_BUILD_IN_DOCKER),y),$(filter-out $(BUILD_OS),$*)))`, but automatic variables like `$*` aren't expanded in GNU make conditionals.  

We might want to just move a lot of this to bash for simplicity and debugging.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -58,32 +52,56 @@ LOCALKUBE_LDFLAGS := "$(K8S_VERSION_LDFLAGS) $(MINIKUBE_LDFLAGS) -s -w -extldfla
...
+	MINIKUBE_BUILD_IN_DOCKER=y
+	LOCALKUBE_BUILD_IN_DOCKER=y
+endif
+
+# If not on linux, localkube must be built in docker
+ifneq ($(BUILD_OS),Linux)
+	LOCALKUBE_BUILD_IN_DOCKER=y
+endif
+
+ifeq ($(IN_DOCKER),1)
```
Comment (dlorenc) [inline Makefile]: Maybe just a quick comment here explaining that this prevents infinite recursion, at first this reads incorrect.

**Comment by dlorenc** [comment]
Nice, this looks really good.

File: Makefile [hunk 2]
```diff
@@ -58,32 +52,59 @@ LOCALKUBE_LDFLAGS := "$(K8S_VERSION_LDFLAGS) $(MINIKUBE_LDFLAGS) -s -w -extldfla
...
+MINIKUBE_ENV_LINUX   := CGO_ENABLED=1 GOARCH=amd64 GOOS=linux
+MINIKUBE_ENV_DARWIN  := CC=o64-clang CXX=o64-clang++ CGO_ENABLED=1 GOARCH=amd64 GOOS=darwin
+MINIKUBE_ENV_WINDOWS := CGO_ENABLED=0 GOARCH=amd64 GOOS=windows
+
+MINIKUBE_DOCKER_CMD := docker run -e IN_DOCKER=1 --user $(shell id -u):$(shell id -g) --workdir /go/src/$(REPOPATH) --entrypoint /bin/bash -v $(PWD):/go/src/$(REPOPATH) $(MINIKUBE_BUILD_IMAGE) -c
+KUBE_CROSS_DOCKER_CMD := docker run -w /go/src/$(REPOPATH) --user $(shell id -u):$(shell id -g) -e IN_DOCKER=1 -v $(shell pwd):/go/src/$(REPOPATH) $(LOCALKUBE_BUILD_IMAGE)
+
+# $(call MINIKUBE_GO_BUILD_CMD, output file, OS)
+define MINIKUBE_GO_BUILD_CMD
+	$($(shell echo MINIKUBE_ENV_$(2) | tr a-z A-Z)) go build --installsuffix cgo -ldflags="$(MINIKUBE_LDFLAGS) $(K8S_VERSION_LDFLAGS)" -o $(1) k8s.io/minikube/cmd/minikube
```
Comment (r2d4) [inline Makefile]: Quick question: I accidently dropped the `-a` flag to go build here, but I don't think we need it.  What was the reasoning initially?
```
        -a
                force rebuilding of packages that are already up-to-date.
```

Comment (r2d4) [inline Makefile]: @dlorenc

Comment (r2d4) [inline Makefile]: added it back

**Comment by r2d4** [comment]
@minikube-bot retest this please

**Pull request merged**